### PR TITLE
feat(targetconnectionmanager): limit concurrent connections, do work async

### PR DESCRIPTION
**Pull request by andrewazores**
Related to #756 

- fix(TargetConnectionManager): limit concurrent connections
- rename variable

Replace previous cache max size behaviour with max concurrent
connections behaviour. This is implemented using a semaphore. When a new
connection is opened a permit is drained from the semaphore. When the
connection closes, whether explicitly, due to connection failure, or due to cache TTL expiry, a
permit is released.

### CODE REVIEW

**Comment by andrewazores** [comment]
### Additional details:

#### Unchanged:

The cache fills when a new JMX connection is opened for a target and drains when that connection is closed. Most/all places in the codebase that obtain a JFRConnection through the TargetConnectionManager in this way intentionally leave the connection open after use so that the TCM may cache it for subsequent requests against the same target. The TCM also has per-target locks to ensure that multiple requests to the same target (or at least, on the same URL[0]) are strictly ordered, ie tasks can't run concurrently on the same target.

The cache drains when a JMX connection is explicitly closed (rarely/never as above), or when the cache entry expires. This expiry is time-based using the TTL parameter. The cache automatically expires and evicts entries TTL seconds after the last time the entry was accessed, ie the last time a task was run on that connection.

#### Changes:

The previous size parameter was used to control the maximum cache size. When set to a negative number the cache would be allowed to grow unbounded (to unlimited size). When set to 0 the cache would be effectively disabled - cached entries would be immediately invalidated and evicted from the cache. When set to a positive number the cache would be limited to this size. If the cache filled to this size and a new connection was requested, *the oldest entry still in the cache would be forcibly evicted*. This entry could potentially even be in use and executing a task.

The new size parameter is renamed to better reflect its behaviour. Instead of forcing an eviction of the oldest cache entry when the cache is full, the new overfilled request must wait until the cache drains before it can be fulfilled. Cache entries can no longer be forcibly evicted by new requests. Cache entries may still be evicted and connections closed while a task is running, but this can only occur if a task is running for longer than the connection TTL. There is a separate method that can be used for tasks that are intentionally potentially long-running, like copying JFR data across the connection, so this should not normally be an issue. The new size parameter has a default value of `-1` which indicates no limit. Any value greater than 0 will enable this "wait for the cache to drain" behaviour.

The TTL parameter is now handled slightly differently. If the TTL is positive the behaviour is unchanged. If the TTL is 0 or negative then the cache is disabled as described earlier, as if the old size parameter were set to 0 - connections are evicted from cache and closed immediately after task execution.

### Context / Motivation

As described above, the previous behaviour allowed (perhaps even encouraged) configurations where JMX connections could be forcibly closed by new connections being opened on a size-limited cache. For deployments with large numbers of target applications this can cause broken API requests and inconsistent/indeterminate behaviour. The only consistently valid configuration was an unbounded cache, but this would allow a Cryostat deployment to potentially open *N* simultaneous connections to *N* target applications, which could be a massive drain on resources.

The new solution allows the cache size parameter to actually be useful in practice, since it will not cause requests to become indeterminate/inconsistent, but only reduces Cryostat API throughput if a large number of requests are being made to different targets, or if an Automated Rule or GraphQL request triggers connections on a large number of targets (where "large" essentially means "greater than cache size").

**Comment by andrewazores** [comment]
Some more changes in the last series of patches.

The TCM cache is now populated by an async job, ie if there is a cache miss then a background thread will process creating the new connection and adding it to the cache. For existing callers of `executeConnectedTask` this means that multiple connections can be opened in parallel, ie a connection does not need to be fully opened before the next one can start opening. This improves throughput, even for callers that block on the result of their connected task.

To alleviate that I have also added `executeConnectedTaskAsync`, which performs the same async cache loading and runs the task async as well, returning a `CompletableFuture` back to the caller. This allows calling code to properly handle tasks async and non-blocking. So far I have only used this within the RuleProcessor.

**Comment by maxcao13** [comment]
Wow, just running this on PR by itself feels really good and doesn't have the same problems as I've mentioned in #1047, it feels like all the issues I've been running into have been resolved. Thanks for doing this! I will review the code next week.

**Review by lkonno — APPROVED** [review]
It is good from my tests. I noticed before some issues during the start up, timeout messages or unresponsiveness, but it seems to be solved with these changes.

File: pom.xml [hunk 1]
```diff
@@ -503,7 +503,7 @@
               <argument>--env</argument>
               <argument>CRYOSTAT_ENABLE_JDP_BROADCAST=true</argument>
               <argument>--env</argument>
-              <argument>CRYOSTAT_TARGET_CACHE_TTL=60</argument>
+              <argument>CRYOSTAT_TARGET_CACHE_TTL=30</argument>
```
[the remark this answers is not in the document]
Comment (andrewazores) [reply pom.xml:506]: That's correct

**Review by maxcao13 — APPROVED** [review]
LGTM!

**Pull request merged**